standalone-metastore/metastore-server/src/main/java/org/apache/hadoop/hive/metastore/HMSHandler.java

```diff
@@ -9037,8 +9061,16 @@ public boolean set_aggr_stats_for(SetPartitionsStatsRequest request) throws TExc
       } else { // No merge.
         Table t = getTable(catName, dbName, tableName);
         // We don't short-circuit on errors here anymore. That can leave acid stats invalid.
-        ret = updatePartitionColStatsInBatch(t, newStatsMap,
-                request.getValidWriteIdList(), request.getWriteId());
+        if (MetastoreConf.getBoolVar(getConf(), ConfVars.TRY_DIRECT_SQL)) {
```

Comment:
It is very strange to use the `ConfVars.TRY_DIRECT_SQL` configuration in client side.

We could have used the same patterns like in `get_table_statistics_req`:
- HMSHandler has only a single method to call on `getMS()`
- ObjectStore could use some `GetHelper` implementation to decide on the actual implementation (directsql/datanucleus)
- If we want to make sure that the size of the Thrift request is not too much we might have another configuration on client size like we do with `BATCH_RETRIEVE_MAX`

Reply:
I think for get operations it's fine. But for set operations we have to start a transaction in HMS handler and for direct sql flow that may cause issue.

Reply:
As discussed offline it would be good to move the batching logic back to the ObjectStore, but that is a different patch altogether.